src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl_SourceGenerators.cs

```diff
@@ -0,0 +1,99 @@
+        {
+            KnownUIContexts.SolutionBuildingContext.UIContextChanged += (_, e) =>
+            {
+                if (!e.Activated)
+                {
+                    // After a build occurs, transition the solution to a new source generator version.  This will
+                    // ensure that any cached SG documents will be re-generated.
+                    this.EnqueueUpdateSourceGeneratorVersion(projectId: null, forceRegeneration: false);
+                }
+            };
+        });
+
+        KnownUIContexts.SolutionExistsAndFullyLoadedContext.WhenActivated(() =>
+        {
+            KnownUIContexts.SolutionExistsAndFullyLoadedContext.UIContextChanged += (_, e) =>
+            {
+                if (e.Activated)
+                {
+                    // After the solution fully loads, transition the solution to a new source generator version.  This
+                    // will ensure that we'll now produce correct SG docs with fully knowledge of all the user's state.
+                    this.EnqueueUpdateSourceGeneratorVersion(projectId: null, forceRegeneration: false);
+                }
+            };
+        });
+
+        // Whenever the workspace status changes, go attempt to update generators.
+        var workspaceStatusService = this.Services.GetRequiredService<IWorkspaceStatusService>();
+        workspaceStatusService.StatusChanged += (_, _) => EnqueueUpdateSourceGeneratorVersion(projectId: null, forceRegeneration: false);
+
+        // Now kick off at least the initial work to run generators.
+        this.EnqueueUpdateSourceGeneratorVersion(projectId: null, forceRegeneration: false);
+    }
+
+    [Export(typeof(ICommandHandler))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [ContentType(ContentTypeNames.XamlContentType)]
+    [Name(PredefinedCommandHandlerNames.SourceGeneratorSave)]
+    [method: ImportingConstructor]
+    [method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    internal partial class SaveCommandHandler() : IChainedCommandHandler<SaveCommandArgs>
+    {
+        public string DisplayName => ServicesVSResources.Roslyn_save_command_handler;
+
+        public CommandState GetCommandState(SaveCommandArgs args, Func<CommandState> nextCommandHandler)
+            => nextCommandHandler();
+
+        public void ExecuteCommand(SaveCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            nextCommandHandler();
+
+            // After a save happens, enqueue a request to run generators on the projects impacted by the save.
+            foreach (var group in args.SubjectBuffer.GetRelatedDocuments().GroupBy(d => d.Project.Solution.Workspace))
+            {
+                if (group.Key is VisualStudioWorkspaceImpl visualStudioWorkspace)
+                {
+                    foreach (var projectGroup in group.GroupBy(d => d.Project))
```

Comment:
>(d => d.Project)) [](http://example.com/codeflow?start=62&length=17)

Could you groupby project on the outer foreach and not need an inner one?

Reply:
(something like)

            foreach (var projectGroup in args.SubjectBuffer.GetRelatedDocuments().GroupBy(d => d.Project))
            {
                if (projectGroup.Key.Solution.Workspace is VisualStudioWorkspaceImpl visualStudioWorkspace)
                {
                    visualStudioWorkspace.EnqueueUpdateSourceGeneratorVersion(projectGroup.Key.Id, forceRegeneration: false);
                }
            }

Reply:
Woorks for me!